### PR TITLE
3.2/feature/validation changes

### PR DESCRIPTION
**Pull request by zeelot**
- #3424 http://dev.kohanaframework.org/issues/3424
- #4062 http://dev.kohanaframework.org/issues/4062
- #4034 http://dev.kohanaframework.org/issues/4034
- #3552 http://dev.kohanaframework.org/issues/3552

As many commits as possible so we can still debate things if anyone complains, but I did confirm all the changes with @zombor before applying :)

### CODE REVIEW

**Comment by zeelot** [comment]
Ah, forgot to make ArrayAccess read-only and unit test the methods... I will take care of this tonight.

File: classes/kohana/validation.php [hunk 1]
```diff
@@ -63,19 +119,31 @@ class Kohana_Validation extends ArrayObject {
...
 	 * @return  array
 	 */
 	public function as_array()
 	{
-		return $this->getArrayCopy();
+		return $this->_data;
+	}
+
+	/**
+	 * Returns the array representation of the current object.
```
Comment (cbandy) [inline classes/kohana/validation.php]: This is not clear enough for me. It should say something about the data/values.

Comment (zeelot) [inline classes/kohana/validation.php]: How about: `Returns the array of data to be validated.`

Comment (cbandy) [inline classes/kohana/validation.php]: I like it.

**Comment by zeelot** [comment]
This should be everything again :)
- reverted back to `_empty_rules`
- reverted all the `Valid` methods to make more sense
- fixed an issue with `Valid::email_domain` running on empty values
- added tests for `Validation` `ArrayAccess` methods
- made `Validation` read-only (`ArrayAccess`)
- added tests for `Valid` methods to run on `NULL` and `FALSE` values

**Pull request merged**